// bootstrap-vue-next/dist/keys-D3gOZjD-.mjs
const withBvnPrefix = (value, suffix = "") => {
  const suffixWithTrail = `${suffix}___`;
  return `___BVN__ID__${value}__${suffix ? suffixWithTrail : ""}`;
};
const createBvnInjectionKey = (name) => withBvnPrefix(name);
const createBvnPluginInjectionKey = (name) => withBvnPrefix(`${name}__plugin`);
const carouselInjectionKey = createBvnInjectionKey("carousel");
const tabsInjectionKey = createBvnInjectionKey("tabs");
const progressInjectionKey = createBvnInjectionKey("progress");
const listGroupInjectionKey = createBvnInjectionKey("listGroup");
const avatarGroupInjectionKey = createBvnInjectionKey("avatarGroup");
const accordionInjectionKey = createBvnInjectionKey("accordion");
const checkboxGroupKey = createBvnInjectionKey("checkboxGroup");
const radioGroupKey = createBvnInjectionKey("radioGroup");
const collapseInjectionKey = createBvnInjectionKey("collapse");
const globalShowHideStorageInjectionKey = createBvnPluginInjectionKey("globalShowHideStorage");
const dropdownInjectionKey = createBvnInjectionKey("dropdown");
const navbarInjectionKey = createBvnInjectionKey("navbar");
const rtlPluginKey = createBvnPluginInjectionKey("rtl");
const breadcrumbPluginKey = createBvnPluginInjectionKey("breadcrumbPlugin");
const modalManagerPluginKey = createBvnPluginInjectionKey("modalManager");
const defaultsKey = createBvnPluginInjectionKey("defaults");
const inputGroupKey = createBvnInjectionKey("inputGroup");
const buttonGroupKey = createBvnInjectionKey("buttonGroup");
const toastPluginKey = createBvnPluginInjectionKey("toast");
const modalControllerPluginKey = createBvnPluginInjectionKey("modalController");
const popoverPluginKey = createBvnPluginInjectionKey("popover");
const formGroupPluginKey = createBvnInjectionKey("formGroupPlugin");
export {
  modalManagerPluginKey as a,
  breadcrumbPluginKey as b,
  collapseInjectionKey as c,
  defaultsKey as d,
  accordionInjectionKey as e,
  avatarGroupInjectionKey as f,
  globalShowHideStorageInjectionKey as g,
  buttonGroupKey as h,
  checkboxGroupKey as i,
  formGroupPluginKey as j,
  radioGroupKey as k,
  inputGroupKey as l,
  modalControllerPluginKey as m,
  listGroupInjectionKey as n,
  progressInjectionKey as o,
  popoverPluginKey as p,
  tabsInjectionKey as q,
  rtlPluginKey as r,
  dropdownInjectionKey as s,
  toastPluginKey as t,
  carouselInjectionKey as u,
  navbarInjectionKey as v,
  withBvnPrefix as w
};
